main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala

```diff
@@ -373,6 +374,137 @@ object Simplifier {
     }
   }
 
+  /** Copy of [[visitType]] that returns [[Type]] instead. */
+  private def visitPolyType(tpe: Type): Type = {
+    val base = tpe.baseType
+    val args = tpe.typeArguments.map(visitPolyType)
+
+    base match {
+      case v@Type.Var(_, _) => Type.mkApply(v, args, tpe.loc)
+
+      case Type.Cst(tc, _) =>
+        tc match {
+          case TypeConstructor.Void => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.AnyType => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.Unit => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.Null => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.Bool => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.Char => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.Float32 => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.Float64 => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.BigDecimal => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.Int8 => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.Int16 => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.Int32 => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.Int64 => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.BigInt => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.Str => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.Regex => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.RecordRowEmpty => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.Sender => throw InternalCompilerException("Unexpected Sender", tpe.loc)
+
+          case TypeConstructor.Receiver => throw InternalCompilerException("Unexpected Receiver", tpe.loc)
+
+          case TypeConstructor.Lazy => Type.mkApply(Type.Cst(tc, tpe.loc), args, tpe.loc)
+
+          case TypeConstructor.Enum(sym, _) => Type.mkApply(Type.Cst(tc, tpe.loc), args, tpe.loc)
+
+          case TypeConstructor.Struct(sym, _) => Type.mkApply(Type.Cst(tc, tpe.loc), args, tpe.loc)
+
+          case TypeConstructor.RestrictableEnum(sym, _) =>
+            val enumSym = new Symbol.EnumSym(sym.namespace, sym.name, sym.loc)
+            Type.mkEnum(enumSym, args, tpe.loc)
+
+          case TypeConstructor.Native(clazz) => Type.Cst(tc, tpe.loc)
+
+          case TypeConstructor.Array =>
+            // remove region
+            // OBS: non-well kinded type
```

Comment:
Yikes, okay

Reply:
perhaps WARNING: This construct a ...

Reply:
I assume the reason is that we don't want to monomorph w.r.t. regions?

Reply:
Monomorphization has already happened. It would still work if left out, but I think the only sane option is to apply the phase logic to both the MonoTypes and the enum/struct types. Otherwise, you have cross-phase invariants, and types become out of sync